File: DanceLib/DanceFilter.cs
using System.Collections.Generic;
using System.Linq;

namespace DanceLibrary
{
    public class DanceFilter(
        IEnumerable<string> styles = null, IEnumerable<string> organizations = null,
        IEnumerable<string> groups = null, Meter meter = null)
    {
        public List<string> Styles { get; } = styles != null ? [.. styles] : [];
        public List<string> Organizations { get; } = organizations != null ? [.. organizations] : [];
        public List<string> Groups { get; } = groups != null ? [.. groups] : [];
        public Meter Meter { get; } = meter;

        public DanceType Reduce(DanceType type)
        {
            if (!MatchMeter(type) || !MatchGroups(type) || !MatchOrganizations(type))
            {
                return null;
            }

            // Get the instances corresponding with the styles, return null if there are none
            var coversOrgs = CoversOrganizations(type);
            var instances = GetMatchingInstances(type)
                .Select(inst => inst.ReduceExceptions(coversOrgs ? null : Organizations))
                .ToList();
            return instances.Count > 0 ? type.Reduce(instances) : null;
        }

        public IEnumerable<DanceType> Filter(IEnumerable<DanceType> types)
        {
            return types.Select(Reduce).Where(type => type != null);
        }

        internal bool MatchMeter(DanceType type)
        {
            return Meter == null || Meter == type.Meter;
        }

        // TODO: These two function could share a common implementation
        internal bool MatchGroups(DanceType type)
        {
            return Groups.Count == 0 || type.Groups.Any(g => Groups.Contains(g.Name));
        }

        internal bool MatchOrganizations(DanceType type)
        {
            return Organizations.Count == 0 || type.Organizations.Any(o => Organizations.Contains(o));
        }

        private List<DanceInstance> GetMatchingInstances(DanceType type)
        {
            return Styles.Count > 0
                ? [.. type.Instances.Where(inst => Styles.Contains(inst.Style))]
                : type.Instances;
        }

        private bool CoversOrganizations(DanceType type)
        {
            return Organizations.Count == 0 || type.Organizations.All(o => Organizations.Contains(o));
        }
    }
}

File: m4d/ClientApp/src/models/DanceDatabase/DanceFilter.ts
import type { DanceInstance } from "./DanceInstance";
import type { DanceType } from "./DanceType";
import type { Meter } from "./Meter";

export class DanceFilter {
  public styles?: string[];
  public organizations?: string[];
  public groups?: string[];
  public meters?: Meter[];

  constructor(init?: Partial<DanceFilter>) {
    this.styles = init?.styles;
    this.organizations = init?.organizations;
    this.groups = init?.groups;
    this.meters = init?.meters;
  }

  public reduce(type: DanceType): DanceType | null {
    if (!this.matchMeter(type) || !this.matchGroups(type) || !this.matchOrganizations(type)) {
      return null;
    }

    const coversOrgs = this.coversOrganizations(type);
    const instances = this.getMatchingInstances(type).map((inst) =>
      inst.reduceExceptions(coversOrgs ? undefined : this.organizations),
    );
    return instances.length > 0 ? type.reduce(instances) : null;
  }

  public filter(types: DanceType[]): DanceType[] {
    // Don't know why the explicit cast is needed here - I guess at some leve the guarantee
    // that the array elements are non-null happens at runtime, but not at compile time?
    return types.map((t) => this.reduce(t)).filter((type) => type !== null) as DanceType[];
  }

  private matchMeter(type: DanceType): boolean {
    return this.meters === undefined || !!this.meters.find((m) => m.equals(type.meter));
  }

  private matchGroups(type: DanceType): boolean {
    return this.groups === undefined || type.groups.some((g) => this.groups!.includes(g.name));
  }

  private matchOrganizations(type: DanceType): boolean {
    return (
      this.organizations === undefined ||
      type.organizations.some((o) => this.organizations!.includes(o))
    );
  }

  private getMatchingInstances(type: DanceType): DanceInstance[] {
    return this.styles !== undefined
      ? type.instances.filter((inst) => this.styles!.includes(inst.style))
      : type.instances;
  }

  private coversOrganizations(type: DanceType): boolean {
    return (
      this.organizations === undefined ||
      type.organizations.every((o) => this.organizations!.includes(o))
    );
  }
}
